File: src/problems/pro/192.kt
package problems.pro

import java.util.Scanner

fun main(){
  val reader = Scanner(System.`in`)
  val matrix = mutableListOf<MutableList<String>>()

  for (i in 0 until 10){
    val row = reader.nextLine().split(" ").toMutableList()
    matrix.add(row)
  }

  var hasOrcMage = false
  for (i in 0 until 10){
    for (j in 0 until 5){
      if (matrix[i][j] == "m"){
        hasOrcMage = true
      }
    }
  }

  var hasHumanMage = false
  for (i in 0 until 10){
    for (j in 5 until 10){
      if (matrix[i][j] == "m"){
        hasHumanMage = true
      }
    }
  }

  for (i in 0 until 10){
    for (j in 0 until 5){
      if (matrix[i][j] == "m"){
        hasOrcMage = true
      }
    }
  }

  val orcPower = orcCalc(matrix, hasOrcMage)
  val humanPower = humanCalc(matrix, hasHumanMage)
  println("orcPower = $orcPower humanPower = $humanPower")
}

private fun orcCalc(matrix: MutableList<MutableList<String>>, hasMage: Boolean): Int{
  var power = 0
  val type = "o"
  for (i in 0 until 10){
    for (j in 0 until 5){
      // UP
      var curr = 0
      if (matrix[i][j] == type || matrix[i][j] == "m"){
        if (i != 0 && matrix[i][j] == type){
          curr++
        }
        // RIGHT
        if (j != 4 && matrix[i][j] == type){
          curr++
        }
        // DOWN
        if (i != 9 && matrix[i][j] == type){
          curr++
        }
        // LEFT
        if (j != 0 && matrix[i][j] == type){
          curr++
        }
        if (curr >= 2 && hasMage){
          power += 3
        } else {
          power++
        }
      }
    }
  }
  return power
}

private fun humanCalc(matrix: MutableList<MutableList<String>>, hasMage: Boolean): Int{
  var power = 0
  val type = "h"
  for (i in 0 until 10){
    for (j in 5 until 10){
      var curr = 0
      // UP
      if (matrix[i][j] == type || matrix[i][j] == "m"){
        if (i != 0 && matrix[i][j] == type){
          curr++
        }
        // RIGHT
        if (j != 9 && matrix[i][j] == type){
          curr++
        }
        // DOWN
        if (i != 9 && matrix[i][j] == type){
          curr++
        }
        // LEFT
        if (j != 5 && matrix[i][j] == type){
          curr++
        }
        if (curr == 4 && hasMage){
          power += 5
        } else {
          power++
        }
      }
    }
  }
  return power
}

File: src/problems/pro/192.py
def calc_power(matrix):
    orc_power = 0
    human_power = 0

    has_human_mage = False
    has_orc_mage = False

    for l in range(10):
        for i in range(0, 5):
            if matrix[l][i] == "m":
                has_orc_mage = True
        for i in range(5, 10):
            if matrix[l][i] == "m":
                has_human_mage = True


    for i in range(10):
        for j in range(0,5):
            if matrix[i][j] == 'o' or matrix[i][j] == 'm':
                orc_curr_power = 1
                if i > 0 and matrix[i - 1][j] == 'o':
                    orc_curr_power += 1
                if i < 9 and matrix[i + 1][j] == 'o':
                    orc_curr_power += 1
                if j > 0 and matrix[i][j - 1] == 'o':
                    orc_curr_power += 1
                if j < 9 and matrix[i][j + 1] == 'o':
                    orc_curr_power += 1
                
                if has_orc_mage:
                    orc_power += orc_curr_power
                else:
                    orc_power += 1

        for j in range(5,10):
            if matrix[i][j] == 'h' or matrix[i][j] == 'm':
                human_curr_power = 1
                if i > 0 and matrix[i - 1][j] == 'h':
                    human_curr_power += 1
                if i < 9 and matrix[i + 1][j] == 'h':
                    human_curr_power += 1
                if j > 0 and matrix[i][j - 1] == 'h':
                    human_curr_power += 1
                if j < 9 and matrix[i][j + 1] == 'h':
                    human_curr_power += 1
                
                if has_human_mage:
                    human_power += human_curr_power
                else:
                    human_power += 1
                

    return orc_power, human_power


def define_winner(matrix):
    orc_power, human_power = calc_power(matrix)

    if orc_power > human_power:
        return "Loktar Ogar!!!"
    elif human_power > orc_power:
        return "Pela Alianca!"
    else:
        return "Batalha lendaria!"

matrix = []
for _ in range(10):
    row = input().strip().split()
    matrix.append(row)

output = define_winner(matrix)
print(output)

